rpc: Pass const char* to evhttp_connection_get_peer for new libevent

Perlover opened the pull request:
The second parameter of evhttp_connection_get_peer in libevent already has type as `const char **`
The compilation of bitcoind with the fresh libevent occurs errors

Details: https://github.com/bitcoin/bitcoin/issues/23606

Perlover committed:
The evhttp_connection_get_peer function from libevent changes the type of the second parameter. Fixing the problem.

Perlover commented:
I tested my compilation with last commit - everything was compiled!
So I think this patch is OK
If you need I can remake this patch to one single commit (in Monday)

hebasto commented:
> The second parameter of evhttp_connection_get_peer in libevent already has type as `const char **` The compilation of bitcoind with the fresh libevent occurs errors

There were no upstream releases for now since the commit https://github.com/libevent/libevent/commit/a18301a2bb160ff7c3ffaf5b7653c39ffe27b385 has been merged, right?

Perlover commented:
@hebasto Yes, everything said is true! I used the master branch of libevent so this commit definitely merged. The old versions have chat ** type for second parameter. So I patched through preprocessor of C++.

Perlover commented:
I rebased these patches to the one patch.

laanwj commented on src/httpserver.cpp:
```diff
@@ -597,7 +601,12 @@ CService HTTPRequest::GetPeer() const
         uint16_t port = 0;
-        evhttp_connection_get_peer(con, (char**)&address, &port);
+
+        #if HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR
```
Please align `#`'s to the left.

laanwj commented on src/httpserver.cpp:
```diff
@@ -597,7 +601,12 @@ CService HTTPRequest::GetPeer() const
-        evhttp_connection_get_peer(con, (char**)&address, &port);
+
+        #if HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR
+        evhttp_connection_get_peer(con, (const char **)&address, &port);
```
Is any case still needed in this case? `&address` *is* `const char**`, right?

laanwj commented:
Please update the PR title to be clearer. "evhttp_connection_get_peer of libevent" is not a good summary of the change.

Perlover commented:
> Please update the PR title to be clearer. "evhttp_connection_get_peer of libevent" is not a good summary of the change.

My English is not native, I tried to give a more precise definition for this PR. I also applied the changes recommended by you and made a rebase relative to the current master branch.

luke-jr commented on configure.ac:
```diff
@@ -1510,6 +1510,13 @@ if test x$build_bitcoin_cli$build_bitcoind$bitcoin_enable_qt$use_tests$use_bench
   fi
 fi
 
+AC_MSG_CHECKING([for evhttp_connection_get_peer])
```
```suggestion
AC_MSG_CHECKING([if evhttp_connection_get_peer expects const char *])
```

luke-jr commented:
Further issues:

1. In this scope, libevent may not even be installed. (Put the check inside the above `if` block that checks for libevent)
2. libevent may require certain CFLAGS to be used. (`CXXFLAGS="$CXXFLAGS $EVENT_CFLAGS"`; don't forget `TEMP_CXXFLAGS` stuff)
3. `uint16_t` needs `<cstdint>` included

Suggested fix for these and prior issues: https://gist.github.com/luke-jr/f8df9d7173a3f673b82b4f0c26854807

Perlover commented on configure.ac:
```diff
@@ -1510,6 +1510,13 @@ if test x$build_bitcoin_cli$build_bitcoind$bitcoin_enable_qt$use_tests$use_bench
 
+AC_MSG_CHECKING([for evhttp_connection_get_peer])
+AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <event2/http.h>]],
+  [[ evhttp_connection_get_peer((evhttp_connection*) nullptr,(const char**) nullptr,(uint16_t *) nullptr); ]])],
```
If I understand correctly, the macro `AC_COMPILE_IF ELSE` only checks the possibility of compiling, not running the program. Therefore, I do not understand why to change the parameters as nullptr to any fictitious ones (as in your final example). By and large, I created the code using examples from the same configuration file as this one, for example:

```
...
AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <ctime>]],
  [[ gmtime_r((const time_t *) nullptr, (struct tm *) nullptr); ]])],
...
```

luke-jr replied:
If libevent some day marks the arguments as `nonnull`, the compiler may error if `nullptr` is passed

Perlover commented:
> Suggested fix for these and prior issues: https://gist.github.com/luke-jr/f8df9d7173a3f673b82b4f0c26854807

```
[ AC_MSG_RESULT([no])]
```

If I understand your code correctly, then there will be no constants defined here and I will have to add additional `#ifdef` or `#if defined(...)` checks to my code.

Perlover commented:
@luke-jr, I applied several of your suggestions - description lines in `configure.ac`, short name commit, added `include <cstdio>`. I also corrected the style of pointers in accordance with the widely used style inside the bitcoin code. At the same time, I have not yet checked for installing libevent (it seems to me that the check is already being performed above and it is superfluous), and I also did not add the `CXXFLAGS` options - to be honest, it seems to me superfluous for a small test function that checks the type of parameters. I am not in favor of complicating things that can not be complicated. In addition, the code passed all automatic checks. I did not replace the parameters with `nullptr` with others, since the macro only checks the possibility of compilation. I don't see the point in that. I used the `nullptr` in a similar way, since this usage is already widely used in `configure.ac`

luke-jr commented:
Right now, your PR will still break on systems:
1. without libevent
2. where libevent is installed outside the compiler's default include path
3. when/if the compiler is some day strict about disallowing nullptr where it is invalid

(Also note you should ideally not rebase unless there are conflicts)

luke-jr commented:
>There is such a code just above my code. Doesn't it guarantee that configure will stop before my code is executed? 

The check above only runs with that `if test x$build_bitcoin_cli$build_bitcoind$bitcoin_enable_qt$use_tests$use_bench != xnonononono; then` conditional. Your code currently runs even if this libevent check is skipped. I am suggesting moving it up inside the `if` block so it only runs when libevent is used.

>Is it really that important to use this complex construction with CXXFLAGS here? All 15 checks performed on github pass.

If libevent is installed outside the default compiler include path, `EVENT_CFLAGS` will contain necessary `-I...` parameters for the compiler.

luke-jr commented:
>If I understand your code correctly, then there will be no constants defined here and I will have to add additional #ifdef or #if defined(...) checks to my code.

This is how other similar checks are currently done.

Perlover committed:
Necessary improvements to make configure work without libevent installed

Perlover commented:
@luke-jr , I did your suggestions in next commit. I hope that they will pass all the tests. Thanks!

luke-jr reviewed (approved):
tACK c62d763fc313585d79ad833c9d729f6acf2652aa

laanwj commented:
Code review ACK c62d763fc313585d79ad833c9d729f6acf2652aa
